feat: record viewable impression

sk- opened the pull request:
Instead of recording the impression when it's seen in the DOM, we record the impression when it's 50% visible. This will only be used in case the IntersectionObserver is available.

PabloReszczynski commented:
It would be nice to test that an impression event doesn't trigger twice when the user scrolls back and forth

sk- commented:
> It would be nice to test that an impression event doesn't trigger twice when the user scrolls back and forth

There was logic to not report the same event twice within the given page, for which I added a test.

And also when the intersection is triggered we stop observing that item. I added a test for that as well.

Note: for some reason I had to change the type of source maps we generate.


@PabloReszczynski one thing though that I realized, is that within a page, unpromoted products will be reported only once. That is, if within the same page you have the same product appearing twice in the page, only one such impressions will be reported.

Also, in single page application, unpromoted products may not be reported when interacted with repeatedly.

For example in this case:

```
page: test.html click on product P
navigate to foo.html
...
navigate to test.html
page: test.html click on Product P
```

Depending on how many events happened in between the two clicks to P, we may not report the second one. We have a queue of 2500 events, and if we get a repeated one, then we don't report it.
